### PR TITLE
xds: fix the validation code to accept new-style CertificateProviderPluginInstance wherever used

### LINKED ISSUE #8885: xds: 'Unsupported configurations in UpstreamTlsContext!' error on creating a channel
### What version of gRPC-Java are you using?

1.44.0, 1.43.2

### What is your environment?

OS Name:                   Microsoft Windows 11 Pro
OS Version:                10.0.22000 N/A Build 22000
openjdk 11.0.8 2020-07-14

### What did you expect to see?

I create a channel using the following function and then use it for communication with a gRPC service as a client.

```
fun createChannelUsingXds(
    name: String
): ManagedChannel {
    val credentials = XdsChannelCredentials.create(InsecureChannelCredentials.create())
    return Grpc.newChannelBuilder("xds:///$name", credentials).build()
}
```

### What did you see instead?

I got an exception:

```
Caused by: java.lang.UnsupportedOperationException: Unsupported configurations in UpstreamTlsContext!

    io.grpc.StatusException: UNKNOWN: Channel Pipeline: [SdsProtocolNegotiators$BufferReadsHandler#0, SdsProtocolNegotiators$ClientSdsHandler#0, WriteBufferingAndExceptionHandler#0, DefaultChannelPipeline$TailContext#0]
        at io.grpc.Status.asException(Status.java:550)
        at io.grpc.kotlin.ClientCalls$rpcImpl$1$1$1.onClose(ClientCalls.kt:295)
        at io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:463)
        at io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:427)
        at io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:460)
        at io.grpc.xds.XdsNameResolver$ConfigSelector$1ClusterSelectionInterceptor$1$1.onClose(XdsNameResolver.java:511)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:562)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:743)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:722)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:920)
     Caused by: java.lang.UnsupportedOperationException: Unsupported configurations in UpstreamTlsContext!
        at io.grpc.xds.internal.sds.ClientSslContextProviderFactory.create(ClientSslContextProviderFactory.java:58)
        at io.grpc.xds.internal.sds.ClientSslContextProviderFactory.create(ClientSslContextProviderFactory.java:27)
        at io.grpc.xds.internal.sds.ReferenceCountingMap.getInternal(ReferenceCountingMap.java:81)
        at io.grpc.xds.internal.sds.ReferenceCountingMap.get(ReferenceCountingMap.java:56)
        at io.grpc.xds.internal.sds.TlsContextManagerImpl.findOrCreateClientSslContextProvider(TlsContextManagerImpl.java:76)
        at io.grpc.xds.internal.sds.SslContextProviderSupplier.getSslContextProvider(SslContextProviderSupplier.java:100)
        at io.grpc.xds.internal.sds.SslContextProviderSupplier.updateSslContext(SslContextProviderSupplier.java:60)
        at io.grpc.xds.internal.sds.SdsProtocolNegotiators$ClientSdsHandler.handlerAdded0(SdsProtocolNegotiators.java:206)
        at io.grpc.netty.shaded.io.grpc.netty.ProtocolNegotiators$ProtocolNegotiationHandler.handlerAdded(ProtocolNegotiators.java:1048)
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:938)
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:609)
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:269)
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:237)
        at io.grpc.netty.shaded.io.grpc.netty.WriteBufferingAndExceptionHandler.handlerAdded(WriteBufferingAndExceptionHandler.java:59)
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:938)
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:609)
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.access$100(DefaultChannelPipeline.java:46)
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1463)
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1115)
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:650)
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:514)
	    at io.grpc.netty.shaded.io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:429)
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486)
        at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
        at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.grpc.netty.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.grpc.netty.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        	... 1 more
```

### Steps to reproduce the bug

We have a golang service which uses gRPC and xDS. It creates an upstream context for a cluster as follows:
image

Go clients work well with this service. I've just started trying to call this service from Java using grpc-xds protocol. 
When I got this error I made some research and found that before creating the `SslContextProvider` for the channel from the `UpstreamContext`, the context is checked for the presence of some fields:

https://github.com/grpc/grpc-java/blob/9f644a086168ebd72e5eabbc5abdafedb2df08c6/xds/src/main/java/io/grpc/xds/internal/sds/ClientSslContextProviderFactory.java#L51-L57

https://github.com/grpc/grpc-java/blob/9f644a086168ebd72e5eabbc5abdafedb2df08c6/xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java#L28-L41

So, the `UpstreamContext` is checked for:
* The filed `TlsCertificateCertificateProviderInstance` has value or
* The field `ValidationContextTypeCase` is `CommonTlsContext.CertificateProviderInstance`  or
* The field `ValidationContextTypeCase` is `CommonTlsContext.CombinedCertificateValidationContext` and the field  `ValidationContextCertificateProviderInstance` has value.

None of these fields is in my  `UpstreamContext`. But after that, some other fields are used for creating the `SslContextProvider`: `CertificateValidationContext.CaCertificatedProviderInstance` and `TlsCertificateProviderInstance`, which are in my context.  So, I would be able to create the `SslContextProvider` if I passed the previous check.

I also noticed that all those fields are `@deprecated` in the [proto](https://github.com/grpc/grpc-java/blob/d4fa0ecc07495097453b0a2848765f076b9e714c/xds/third_party/envoy/src/main/proto/envoy/extensions/transport_sockets/tls/v3/tls.proto#L228) and generated stubs, so I assumed that it could be a bug.

### PR DESCRIPTION
**Pull request by sanjaypujare**
Update `CommonTlsContextUtil` validation functions to accept new-style `CertificateProviderPluginInstance`

fixes #8885

### CODE REVIEW

**Comment by sanjaypujare** [comment]
The fix in this PR verified as reported in https://github.com/grpc/grpc-java/issues/8885#issuecomment-1031418766

File: xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java [hunk 1]
```diff
@@ -37,7 +37,18 @@ private static boolean hasCertProviderValidationContext(CommonTlsContext commonT
...
+
+  private static boolean hasIdentityCertificateProviderInstance(CommonTlsContext commonTlsContext) {
+    return commonTlsContext.hasTlsCertificateProviderInstance()
+        || commonTlsContext.hasTlsCertificateCertificateProviderInstance();
+  }
+
+  private static boolean hasValidationProviderInstance(CommonTlsContext commonTlsContext) {
+    return (commonTlsContext.hasValidationContext() && commonTlsContext.getValidationContext()
+        .hasCaCertificateProviderInstance())
+        || commonTlsContext.hasValidationContextCertificateProviderInstance();
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java:51]: If this method `hasValidationProviderInstance` is only used once, why not use it inline?

File: xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java [hunk 2]
```diff
@@ -27,8 +27,8 @@ private CommonTlsContextUtil() {}
 
   static boolean hasCertProviderInstance(CommonTlsContext commonTlsContext) {
     return commonTlsContext != null
-        && (commonTlsContext.hasTlsCertificateCertificateProviderInstance()
-            || hasCertProviderValidationContext(commonTlsContext));
+        && (hasIdentityCertificateProviderInstance(commonTlsContext)
+        || hasCertProviderValidationContext(commonTlsContext));
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java:31]: What about
```java
if (commonTlsContext == null) {
  return false;
}
return hasIdentityCertificateProviderInstance(commonTlsContext)
    || hasCertProviderValidationContext(commonTlsContext);
```

to avoid `x && (y || z)`? Ditto to avoid `(x && y) || z`.

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java:31]: > ...
> to avoid `x && (y || z)`? Ditto to avoid `(x && y) || z`.

Any particular reason to avoid such expressions? Not aware of any guidelines suggesting that - although the change is easy to make

File: xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java (hunk 1 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java:51]: Similar to other functions defined here e.g. `hasIdentityCertificateProviderInstance` or `hasCertProviderValidationContext`. If I inline all functions used only once I'll end up with complex functions (with successive such inlining). In general if it helps in understanding the code I think it is good to have separate, appropriately named functions. The compiler will anyway inline such functions so I don't see the downside.

File: xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java (hunk 2 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java:31]: > Any particular reason to avoid such expressions?

I'm feeling the expression is too long when `x,y,z` are long and have parentheses, and therefore hard to parse by human.

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/internal/sds/CommonTlsContextUtil.java]: done

**Review by dapengzhang0 — APPROVED** [review]
